src/inference/smc.js

```diff
@@ -145,7 +159,10 @@ module.exports = function(env) {
     var logW = util.logsumexp(_.map(particles, 'logWeight'));
     var logAvgW = logW - Math.log(m);
 
-    assert.notStrictEqual(logAvgW, -Infinity, 'All particles have zero weight.');
+    if (logAvgW === -Infinity) {
+      // do not return, execution continues
+      env.coroutine.error('All particles have zero weight.');
+    }
```

Comment:
Also, I just noticed that because we're not doing `return env.coroutine.error()` here, this won't do the right thing when we're not throwing errors. Inference will (attempt to) continue, rather than returning early with an `Error` instance. For example, this generates an error as expected:

```js
display(Infer({
  method: 'SMC',
  throwOnError: true,
  particles: 2,
  model: function() {
    condition(false);
  }
}));
```

but with `{throwOnError: false}` it returns a (incorrect) marginal, not an `Error`.

This is similar to the situation we found ourselves in with `getSupport`, and similar solutions will apply. I guess that switching to CPS will again be the least worst option.